### PR TITLE
PMM-10231 Service Params Source

### PR DESCRIPTION
**Pull request by pkadej**
[PMM-10231](https://jira.percona.com/browse/PMM-10231)

Build: SUBMODULES-2719

### CODE REVIEW

File: agent/credentialssource/credentials_source.go
Comment (BupycHuk) [inline agent/credentialssource/credentials_source.go:102]: Maybe it's better to wrap `err` returned by `parseJsonFile`

File: agent/credentialssource/credentials_source.go
Comment (BupycHuk) [inline agent/credentialssource/credentials_source.go:110]: let's use `errors.Wrapf` in all cases below with `%w`

File: agent/credentialssource/credentials_source.go
Comment (BupycHuk) [inline agent/credentialssource/credentials_source.go:106]: no reason to initialize empty strings.

File: agent/credentialssource/credentials_source.go
Comment (BupycHuk) [inline agent/credentialssource/credentials_source.go:115]: I still think that we don't need to check the file permissions.

File: agent/credentialssource/credentials_source.go
Comment (BupycHuk) [inline agent/credentialssource/credentials_source.go:136]: if we validate data in both functions can we do it in https://github.com/percona/pmm/pull/1110/files#diff-3f486b304421a573b79690cddeb7fe149e9d4159999e8ad5bad717bbd4d85bc9R60 once we get the result?

File: api-tests/management/mysql_test.go
Comment (BupycHuk) [inline api-tests/management/mysql_test.go:598]: it should be allowed only if `credentials source` is passed. In other cases behaviour should be the same.

File: api-tests/management/mysql_test.go
Comment (BupycHuk) [inline api-tests/management/mysql_test.go:618]: let's revert this change

File: api-tests/management/mysql_test.go
Comment (BupycHuk) [reply api-tests/management/mysql_test.go:598]: the same in all cases below

File: managed/services/agents/credentials_source_agent_invoker.go
Comment (BupycHuk) [inline managed/services/agents/credentials_source_agent_invoker.go:31]: IMHO, it's already in `agents` directory, so `AgentInvoker` suffix is unnecessary. Maybe `CredentialsSourceLoader` or checker?

File: managed/services/agents/credentials_source_agent_invoker.go
Comment (BupycHuk) [inline managed/services/agents/credentials_source_agent_invoker.go:43]: let's rename it to `Credentials`, `GetCredentials` or `LoadCredentials`

File: managed/services/agents/credentials_source_agent_invoker.go
Comment (BupycHuk) [inline managed/services/agents/credentials_source_agent_invoker.go:88]: as far as I see here only service type is different here, so we could use the same mechanism like we use in https://github.com/percona/pmm/blob/4c0aaa7cea4444ae37414ce2c43b21fd1c4772b1/managed/services/management/service.go#L31 to convert from `model` to `inventorypb`

File: managed/services/management/external.go
Comment (BupycHuk) [inline managed/services/management/external.go:74]: let's add additional checks for the field we removed validation

Comment (BupycHuk) [reply managed/services/management/external.go:74]: it should be implemented for 2 cases:
1. credentials source is not passed
2. credentials source is passed, but fields aren't presented in the source.

**Review by BupycHuk — CHANGES_REQUESTED** [review]
Let's store credentials source in database for the future usage.

File: agent/credentialssource/credentials_source.go
Comment (pkadej) [reply agent/credentialssource/credentials_source.go:115]: I just moved whole logic/method from pmm-admin: https://github.com/percona/pmm/blob/main/admin/commands/base.go#L102
Removing it.

File: managed/services/management/mongodb.go
Comment (pkadej) [reply managed/services/management/mongodb.go]: Fixed.

File: managed/services/agents/credentials_source_agent_invoker.go
Comment (pkadej) [reply managed/services/agents/credentials_source_agent_invoker.go]: Changed.

File: agent/credentialssource/credentials_source.go
Comment (pkadej) [reply agent/credentialssource/credentials_source.go]: Changed.

File: agent/credentialssource/credentials_source.go
Comment (pkadej) [reply agent/credentialssource/credentials_source.go]: Done

File: agent/credentialssource/credentials_source.go
Comment (pkadej) [reply agent/credentialssource/credentials_source.go]: Changed.

File: agent/credentialssource/credentials_source.go
Comment (pkadej) [reply agent/credentialssource/credentials_source.go]: Changed.

File: agent/credentialssource/credentials_source.go
Comment (pkadej) [reply agent/credentialssource/credentials_source.go]: Changed.

File: agent/credentialssource/credentials_source.go
Comment (pkadej) [reply agent/credentialssource/credentials_source.go]: @cezmunsta It shouldn't be removed right? Can we decide if should it be keeped or not?

File: managed/services/agents/credentials_source_agent_invoker.go
Comment (pkadej) [reply managed/services/agents/credentials_source_agent_invoker.go]: Done.

File: managed/services/agents/credentials_source_agent_invoker.go
Comment (pkadej) [reply managed/services/agents/credentials_source_agent_invoker.go]: Done.

File: managed/services/management/external.go
Comment (pkadej) [reply managed/services/management/external.go:74]: 1. If credentials source is not passed, pmm-admin sets & sends default username.
2. Added condition: 
check if source-credentials is passed and if no username were found throws the same message as in case before removing validation (it allow us to keep backward compatibility without test/api test changes).

File: api-tests/management/mysql_test.go
Comment (pkadej) [reply api-tests/management/mysql_test.go]: Changed.

File: api-tests/management/mysql_test.go
Comment (pkadej) [reply api-tests/management/mysql_test.go]: Done.

File: managed/services/agents/credentials_source_agent_invoker.go
Comment (pkadej) [reply managed/services/agents/credentials_source_agent_invoker.go]: Changed to `CredentialsSourceLoader`

File: managed/services/agents/credentials_source_loader.go
Comment (BupycHuk) [inline managed/services/agents/credentials_source_loader.go:97]: it's a common practice in go to use `ok` as a second argument to check if the key exists
```suggestion
	inventorypbServiceType, ok := serviceTypes[serviceType]
```

File: managed/services/management/external.go
Comment (BupycHuk) [reply managed/services/management/external.go:74]: API can be used directly, so we shouldn't rely on the `pmm-admin` logic.

Comment (pkadej) [reply managed/services/management/external.go:74]: Got it.
I changed in every place where the field "required" was removed from API:
- proxysql
- postgresql
- mysql

**Review by BupycHuk — APPROVED** [review]

**Comment by pkadej** [comment]
I had to regenerate descriptor.bin, because unused "ParseDefaultsFile" fields in agent API  was renamed.
```
bin/buf breaking --against descriptor.bin api
api/agentpb/agent.proto:1:1:Previously present message "ParseDefaultsFileRequest" was deleted from file.
api/agentpb/agent.proto:1:1:Previously present message "ParseDefaultsFileResponse" was deleted from file.
api/agentpb/agent.proto:694:5:Field "20" with name "parse_credentials_source" on message "AgentMessage" changed option "json_name" from "parseDefaultsFile" to "parseCredentialsSource".
api/agentpb/agent.proto:694:5:Field "20" on message "AgentMessage" changed type from "agent.ParseDefaultsFileResponse" to "agent.ParseCredentialsSourceResponse".
api/agentpb/agent.proto:694:36:Field "20" on message "AgentMessage" changed name from "parse_defaults_file" to "parse_credentials_source".
api/agentpb/agent.proto:726:5:Field "18" with name "parse_credentials_source" on message "ServerMessage" changed option "json_name" from "parseDefaultsFile" to "parseCredentialsSource".
api/agentpb/agent.proto:726:5:Field "18" on message "ServerMessage" changed type from "agent.ParseDefaultsFileRequest" to "agent.ParseCredentialsSourceRequest".
api/agentpb/agent.proto:726:35:Field "18" on message "ServerMessage" changed name from "parse_defaults_file" to "parse_credentials_source".
```

File: admin/commands/management/add_proxysql.go
Comment (michal-kralik) [inline admin/commands/management/add_proxysql.go:118]: Is there a specific reason we replace an empty password or username if not provided?
I could imagine if both are empty, we apply defaults, but password could be an empty string, right?

File: admin/commands/management/add_mysql.go
Comment (michal-kralik) [inline admin/commands/management/add_mysql.go:100]: Can we please improve the help text here and in other instances of `CredentialsSource`?
It's not clear what value is expected here.

**Review by michal-kralik — CHANGES_REQUESTED** [review]

File: admin/commands/management/add_proxysql.go
Comment (pkadej) [reply admin/commands/management/add_proxysql.go:118]: Default credentials for both username and password were "admin" (line 52 - 53), so if `--credentials-source` parameter wasn't passed and `--password` also, pmm-admin should apply 'admin' as default.

Comment (michal-kralik) [reply admin/commands/management/add_proxysql.go:118]: That's correct. There was one case though which we don't support now.

```
pmm-admin add --username=root --password=""
```

This allows you to configure an empty password while the new solution does support this.
That is, of course, unless we did not discover a bug in Kong where the command would use a default value instead :)

Comment (pkadej) [reply admin/commands/management/add_proxysql.go:118]: Got your point, thank you.

If we run pmm-admin from main branch: 
```
pmm-admin add proxysql --username=root --password=""
```
pmm-server will get username=root and password="" in request.
For now pmm-admin allows sending empty password even if default value for proxysql is `admin`.

In this PR even if we pass or not empty parmeter `--password=""` the pmm-server will get default `admin` password in request from admin. 






That's interesting case.
Should pmm-admin support passing empty parameters/passwords even if defaults values has been declared?
CC @BupycHuk

Comment (pkadej) [reply admin/commands/management/add_proxysql.go:118]: > Yeah, it should support it, another question is can we differentiate the flag which wasn't passed from the flag which is empty for reason?

Unfortunately, I don't have idea how we can distinguish them :disappointed:

Comment (michal-kralik) [reply admin/commands/management/add_proxysql.go:118]: It appears we shall be able to distinguish this through [Value.Set](https://pkg.go.dev/github.com/alecthomas/kong#Value) attribute.

Comment (pkadej) [reply admin/commands/management/add_proxysql.go]: Fixed. 
image

File: admin/commands/management/add_mysql.go
Comment (pkadej) [reply admin/commands/management/add_mysql.go]: Done.

File: admin/commands/management/add.go
Comment (BupycHuk) [inline admin/commands/management/add.go:123]: seems like these new types can be easily replaced by just pointers https://github.com/alecthomas/kong#pointers

**Review by BupycHuk — CHANGES_REQUESTED** [review]

File: admin/commands/management/add.go
Comment (pkadej) [reply admin/commands/management/add.go:123]: This is going to be a quite long comment.

When decided to use pointers https://github.com/alecthomas/kong#pointers, pmm-admin compiled but failed to launch:
`panic: AddProxySQLCommand.Username: unsupported field type *string, perhaps missing a cmd:"" tag?`
since --username/--password are **flags** adding cmd:"" will make parameter as a **subcommand**.


I did some research and found simillar problem with pointers in subcommands: https://github.com/alecthomas/kong/issues/196. Issue was solved in latest version which pmm-admin uses, so I jumped into PRs source https://github.com/alecthomas/kong/commit/43616c6a6d5989dd9b3512e80f00ed3337b8b842#diff-679d4747ec9379321e0f2e758c26439f99115a71b4618af9713a65f65d160fb7R1286
The problem is that the fix is only for subcommands, not for flags.

I've replaced

> Cmd *cmd \`cmd:""\`

with
> Cmd   *string \`help:"PostgreSQL username"\`

and the result was the same, as for pmm-admin.


It looks like Kong's **inner commands** have problems with **pointers for parameters**.

In next step i've tried to add `cmd:""` just to see how the app will behave.
Unfortunately, it doesn't work as expected. Passed empty parameter `--username` and skipped `--password` wasn't nill but equal "" in runtime. 
So pointers won't solve this issue :( 



Should we stick to the current solution?

Comment (BupycHuk) [reply admin/commands/management/add.go:123]: Yes let's stick on current solution with some TODO that ot should be fixed once kong will start supporting it. And can we create an issues in kongs repository, please.

Comment (pkadej) [reply admin/commands/management/add.go:123]: Just checked that main branch with kong source code and it works fine. 
then find out that this code https://github.com/alecthomas/kong/pull/296 hasn't been released yet (merged 20th Sept).

There is no need to create new issue. Will add TODO that it should be fixed after kong version bumping.
Thanks!

Comment (BupycHuk) [reply admin/commands/management/add.go]: Great, thank you

**Comment by BupycHuk** [comment]
Feel free to merge without waiting for my approval, just make sure that it was QA-ed

File: admin/commands/management/add.go
Comment (michal-kralik) [reply admin/commands/management/add.go]: @pkadej pointers don't work because they have not been release yet and are not part of v0.6.1
With the master branch they work as expected.

I'd opt for using the pointers. We can either wait until a new release is available or we can pin the module to a specific commit and upgrade as soon as the new version is out.
Either way, I tried pinging the maintainers to see when a new release is due https://github.com/alecthomas/kong/pull/296#issuecomment-1290048303

Comment (michal-kralik) [reply admin/commands/management/add.go]: v0.7.0 has just been released. We shall upgrade and use pointers.

Comment (pkadej) [reply admin/commands/management/add.go]: Done. Updated, retested and pushed. Works. Thanks :)

File: admin/commands/management/add_external.go
Comment (michal-kralik) [inline admin/commands/management/add_external.go:60]: Let's please update the help here so it's clear what is expected.
With the current help it's not clear what the file shall contain.

**Pull request closed without merging**
